File: experiments/monitor-api/src/entities/pod.ts
import { ContainerCategory } from "../models/category";
import { Metric } from '../models/evaluation';
import Queue from './queue';

const TIME_WEIGHT = parseFloat(process.env.TIME_WEIGHT || "100.0")
const METRIC_WEIGHT = parseFloat(process.env.METRIC_WEIGHT || "20.0")
const PROGRESSING_THREESHOLD = parseFloat(process.env.PROGRESSING_THREESHOLD || "10")
const CONVERGED_THREESHOLD = parseFloat(process.env.CONVERGED_THREESHOLD || "-25")


interface ModelEvaluation {
  metric: Metric
  time: number
}

export class Pod {
  name: string
  score: number
  accessedAt: number
  category: ContainerCategory
  evaluations: Queue<ModelEvaluation>

  constructor(name: string) {
    this.name = name
    this.score = 0
    this.accessedAt = Date.now()
    this.category = ContainerCategory.Progressing
    this.evaluations = new Queue<ModelEvaluation>()
  }

  speculate() {
    let size = this.evaluations.size
    if (size < 2) {
      return;
    }

    let oldEvaluation = this.evaluations.seek()!
    while (--size) {
      oldEvaluation = this.evaluations.pop()!
    }

    let newEvaluation = this.evaluations.seek()!
    let dt = newEvaluation.time - oldEvaluation.time
    let dmetric = newEvaluation.metric.value - oldEvaluation.metric.value
    let score = (dmetric * METRIC_WEIGHT - dt / TIME_WEIGHT)

    let oldCategory = this.category
    let dscore = score - this.score
    console.log("time", dt)
    console.log("SCORE", score, dscore)
    if (dscore > PROGRESSING_THREESHOLD || dscore > 5) {
      this.category = ContainerCategory.Progressing
    } else if (dscore < -5) {
      if (this.category == ContainerCategory.Progressing && dscore >= CONVERGED_THREESHOLD) {
        this.category = ContainerCategory.Watching
      } else {
        this.category = ContainerCategory.Converged
      }
    }
    if (oldCategory != this.category) {
      this.score = score
    }
  }

  addEvaluation(evaluation: ModelEvaluation) {
    this.evaluations.push(evaluation)
  }
}

export class PodStorage {
  private pods: { [key: string]: Pod }

  constructor() {
    this.pods = {}
  }

  findPod(name: string) {
    let pod = this.pods[name]
    if (pod) {
      pod.accessedAt = Date.now()
    }
    return pod
  }

  addPod(name: string) {
    const pod = new Pod(name)
    this.pods[name] = pod
    return pod
  }

  garbageCollection() {
    let time = Date.now()
    let pods = { ...this.pods }
    for (let pod in pods) {
      if ((time - pods[pod].accessedAt) >= 3600000) { // 1 hour period
        delete this.pods[pod]
      }
    }
  }
}

const podStorage = new PodStorage()

export default podStorage;

File: experiments/monitor-api-v2/pkg/model/pod.go
package model

import (
	"fmt"
	"log"
	"math"
	"sync"
	"time"

	"github.com/oleiade/lane"
	"github.com/sajari/regression"
)

type Pod struct {
	Name          string
	Score         float64
	CreatedAt     time.Time
	LastMigration time.Time
	AccessedAt    time.Time
	Category      PodCategory
	Evaluations   *lane.Queue
	Regressor     *regression.Regression
	mu            sync.Mutex
}

func NewPod(name string) *Pod {
	t := time.Now()
	r := new(regression.Regression)
	r.SetObserved("log y")
	r.SetVar(0, "t")
	return &Pod{
		Name:        name,
		Score:       0,
		CreatedAt:   t,
		AccessedAt:  t,
		Category:    Progressing,
		Evaluations: lane.NewQueue(),
		Regressor:   r,
	}
}

func (pod *Pod) Speculate() {
	// Speculate if there are 2 evaluation results
	n := pod.Evaluations.Size()
	if n < 2 {
		return
	}

	// Only speculate using latest two
	pod.mu.Lock()
	for ; n > 2; n-- {
		pod.Evaluations.Pop()
	}
	eval1 := pod.Evaluations.Pop().(*PodEvaluation)
	eval2 := pod.Evaluations.First().(*PodEvaluation)
	pod.mu.Unlock()

	// Find differences
	t1 := float64(eval1.Time.UnixMilli())
	t2 := float64(eval2.Time.UnixMilli())
	t0 := float64(pod.CreatedAt.UnixMilli())
	// dt := float64((newest.Time.UnixNano() + oldest.Time.UnixNano()) / 1000000)

	// Get score
	score := math.Abs((eval2.Metric - eval1.Metric) / (t2 - t1))
	alpha := float64(0.0)
	beta := float64(0.0)

	pod.Regressor.Train(regression.DataPoint(math.Log(score)/math.Log(math.E), []float64{t2 - t0}))
	if err := pod.Regressor.Run(); err == nil {
		alpha, err = pod.Regressor.Predict([]float64{t2 - t0})
		if err != nil {
			log.Fatalln(err)
			return
		}
		alpha = math.Exp(alpha)
		fmt.Println(alpha)
		beta = 1 - pod.Regressor.R2
	}

	// log.Printf("Time: %v, Score: %v, dScore: %v", dt, score, dscore)
	// log.Printf("Test %v", utils.PROGRESSING_THREESHOLD)
	// SpeCon + Custom categorization
	if score > alpha {
		pod.Category = Progressing
	} else {
		if pod.Category == Progressing {
			pod.Category = Watching
		} else if score+score*beta < pod.Score {
			pod.Category = Converged
		}
	}
	pod.Score = score
}

func (pod *Pod) AddEvaluation(eval *EvaluationObject) {
	pod.mu.Lock()
	pod.Evaluations.Enqueue(&PodEvaluation{Metric: eval.Value, Time: time.Now()})
	pod.mu.Unlock()
}
